pkg/controller/machine/machine_controller.go

```diff
@@ -278,6 +286,34 @@ func (r *ReconcileMachine) isDeleteAllowed(machine *clusterv1.Machine) bool {
 	return node.UID != machine.Status.NodeRef.UID
 }
 
+// isDeleteNodeAllowed returns true only if the Machine's NodeRef is not nil
+// and if the Machine is not the last control plane node in the cluster.
+func (r *ReconcileMachine) isDeleteNodeAllowed(ctx context.Context, machine *clusterv1.Machine) (bool, error) {
+
+	// Cannot delete something that doesn't exist.
+	if machine.Status.NodeRef == nil {
+		return false, nil
+	}
+
+	// Get all of the machines that belong to this cluster.
+	machines, err := r.getMachines(ctx, machine)
+	if err != nil {
+		return false, err
+	}
+
+	controlPlaneMachines := util.GetControlPlaneMachines(machines)
+
+	// If this machine is a control plane node then only delete the NodeRef
+	// if there is another control plane node (besides this machine) left.
+	if util.IsControlPlaneMachine(machine) {
+		return len(controlPlaneMachines) > 1, nil
```

Comment:
There is likely also going to be an issue with deleting nodes in the case that the user starts with a 3 node controlplane and deletes 2 of the controlplane machines, since etcd will no longer have quorum.

Reply:
Do any of the providers handle removing members from etcd today?

Reply:
Hmm. I don't know enough here to respond to this. Based solely on things from the Kubernetes angle, if the API server is there to respond, it seems the call should occur.

Reply:
@ncdc I'm not aware of any, which is why it was called out in the upgrade doc.

Reply:
Can we proceed with this PR despite the lack of etcd member removal? That will have to come in a control plane manager of some sort, I'd say.

Reply:
Sure, lets proceed.